### ISSUE TITLE
cilium significant increase in startup time from 1.13 to 1.14

### ISSUE THREAD

**Issue by rcanderson23** [cilium/cilium#29795]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We upgraded all our clusters from 1.13 to 1.14.4 and ran into issues where the upgrade timed out (our kube apps are managed by FluxCD and had a 15m timeout along with failing startup probes) causing a rollout. Investigating, we are seeing Cilium pods taking ~5m+ for the startup probe to pass. We have cilium configured to used `etcd` and are in a clustermesh config with a mesh totaling 6 clusters with qps set to 100.

Cluster1 size (~5m+ startup):
```
nodes: ~70
ciliumidentities: ~1800
network policies: ~2500
pods: ~4700
services: ~6100
```

Cluster2 size (~80-90s startup):
```
nodes: ~45
ciliumidentities: ~800
network policies: ~1300
pods: ~2300
services: ~6100
```

Checking our logs `rate({pod="cilium-shtqb"} |= "error"[1m])` there is about 3 errors/sec for the first minute. Most log messages match this pattern:
```

2023-12-11 08:15:36.636 | level=error msg="endpoint regeneration failed" containerID=a0f2666bd0 datapathPolicyRevision=0 desiredPolicyRevision=2573 endpointID=3746 error="getting interface enicc1635990d4 by name: Link not found" identity=103145 ipv4=10.17.88.154 ipv6="fe80::6894:cfff:fecf:5621" k8sPodName=external-secrets/external-secrets-65cf896d4-g2mlr subsys=endpoint |  
2023-12-11 08:15:36.636 | level=warning msg="Regeneration of endpoint failed" bpfCompilation=0s bpfLoadProg="86.599µs" bpfWaitForELF="2.146µs" bpfWriteELF="409.059µs" containerID=a0f2666bd0 datapathPolicyRevision=0 desiredPolicyRevision=2573 endpointID=3746 error="getting interface enicc1635990d4 by name: Link not found" identity=103145 ipv4=10.17.88.154 ipv6="fe80::6894:cfff:fecf:5621" k8sPodName=external-secrets/external-secrets-65cf896d4-g2mlr mapSync="2.256µs" policyCalculation="11.98µs" prepareBuild="267.168µs" proxyConfiguration="10.843µs" proxyPolicyCalculation="12.454µs" proxyWaitForAck=0s reason="retrying regeneration" subsys=endpoint total=2.149403ms waitingForCTClean=271ns waitingForLock=622ns |  
2023-12-11 08:15:36.636 | level=info msg="generating BPF for endpoint failed, keeping stale directory" containerID=a0f2666bd0 datapathPolicyRevision=0 desiredPolicyRevision=2573 endpointID=3746 error="getting interface enicc1635990d4 by name: Link not found" file-path=3746_next_fail identity=103145 ipv4=10.17.88.154 ipv6="fe80::6894:cfff:fecf:5621" k8sPodName=external-secrets/external-secrets-65cf896d4-g2mlr subsys=endpoint
```

I don't have hard data but I do know the rollback happened fast because v1.13 pods were passing startup and becoming ready within seconds of starting compared to the v1.14. Closest issue I have found is https://github.com/cilium/cilium/issues/24987

Any advice to improving startup time or perhaps this is a regression? My biggest concern with startup time taking this long is our upgrade process. We generally have enough capacity so not _too_ worried about scale up time with new nodes.

### Cilium Version

Client: 1.14.4 87dd2b64 2023-11-09T15:02:52-05:00 go version go1.20.11 linux/amd64
Daemon: 1.14.4 87dd2b64 2023-11-09T15:02:52-05:00 go version go1.20.11 linux/amd64

### Kernel Version

Linux ip-10-17-69-121.us-west-2.compute.internal 5.10.198-187.748.amzn2.x86_64 #1 SMP Tue Oct 24 19:49:54 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Server Version: version.Info{Major:"1", Minor:"24+", GitVersion:"v1.24.17-eks-4f4795d", GitCommit:"af19e454a15b5eb16d9f29d4d2361b3050ac78a6", GitTreeState:"clean", BuildDate:"2023-10-20T23:22:36Z", GoVersion:"go1.20.10", Compiler:"gc", Platform:"linux/amd64"}

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by rcanderson23** [comment]
Our cilium agent startup probes are now taking longer than 9m to pass. 

Cluster stats:
```
nodes: ~110
pods: ~6300
network policies: ~3400
services: ~7600
cilium identities: ~1800
```

**Comment by squeed** [comment]
Thanks for the report! It's not good that restarts take so long. I'll see if we have any resources to dig in to policy ingestion time.

It would be good to get some hard data for the bootstrap timing. Can you check out the metrics `cilium_agent_bootstrap_seconds_sum` -- there is a label for every phase of bootstrap. Let's see what, exactly, is the issue.

**Comment by marseel** [comment]
Would be great if you could provide sysdumps from both clusters. The original issue with startup time of 5 minutes might be related to https://github.com/cilium/cilium/issues/26462
tldr now we wait for synchronization of data from remote clusters before marking cilium-agent as ready to prevent connection disruptions.

**Comment by rcanderson23** [comment]
Thanks for the replay squeed and marseel.

Here is the `cilium_agent_bootstrap_seconds_sum` for one of our pods in cluster1 and cluster2

Cluster1:
<img width="1656" alt="image" src="https://github.com/cilium/cilium/assets/15879161/967ab2c9-4355-4ff3-8c8f-0a0b0f472276">

Cluster2:
<img width="1656" alt="image" src="https://github.com/cilium/cilium/assets/15879161/93478047-16bf-47da-a997-6bacd7c18f44">

We have upgraded to v1.14.5 since opening this issue.

> tldr now we wait for synchronization of data from remote clusters before marking cilium-agent as ready to prevent connection disruptions.

This thought definitely crossed my mind but thought the startup time would be more uniform across all clusters in the mesh. 🤷 

Running the sysdumps but it contains information that I would rather not share on the public net. Is there more targeted information that I can get you? Looking at this sysdump it doesn't contain logs from startup as they have already rotated out. Would it be more informative if I restarted the pod and captured the sysdump after startup is successful

**Comment by julianwiedmann** [comment]
so it's pretty much all `k8sInit`. @marseel does that match up with your suspicion re https://github.com/cilium/cilium/issues/26462 ?

**Comment by marseel** [comment]
I was mistaken about the waiting time for clustermesh, it's 1 minute in the worst case.

k8sInit + `DeltaFIFO Pop Process` indicates that indeed startupProbe is blocked on importing all network policies - this is intended because new pods can't be started before network policies are imported. 
This seems like a regression in the latency of importing policies.

**Comment by rcanderson23** [comment]
> Could you provide an exact version of 1.13 that you are upgrading from? This will help us understand what are potential areas to investigate.

We were on v1.13.4 and then upgraded 1.14.4

**Comment by squeed** [comment]
Thanks for the analysis, @marseel. We know of a few bottlenecks within policy import time, but I wasn’t aware it was quite this bad.

Right now @gandro has been working in this area; I’ll assign this to them to take a look as work progresses.

As always, anyone else is also welcome to take a gander :).

**Comment by gandro** [comment]
Unassigning myself, as wasn't able to find any obvious issues. I will also shift away focus from policy for the near future.

It would be interesting to know if the issue still persists. There have been changes to endpoint regeneration since.

**Comment by giorio94** [comment]
Removing the `area/clustermesh` label, as unrelated per the above conversation.

**Comment by julianwiedmann** [comment]
> It would be interesting to know if the issue still persists. There have been changes to endpoint regeneration since.

agree, I think this would benefit from a re-test on v1.16.

**Comment by lrouquette** [comment]
Facing the same issue here, but upgrading from 1.13.16 to 1.14.11 (no cluster-mesh). The policy import seems to be the component taking much longer in 1.14.

**Comment by squeed** [comment]
There are significant improvements to policy import time in v1.16, please consider upgrading.

**Comment by lrouquette** [comment]
> There are significant improvements to policy import time in v1.16, please consider upgrading.

Yes, if it was [officially supported](https://docs.cilium.io/en/v1.16/operations/upgrade/#step-1-upgrade-to-latest-patch-version), I would upgrade from 1.13 to 1.16 directly...

**Comment by joestringer** [comment]
Looks like this is now fixed in https://github.com/cilium/cilium/pull/32703 .
